### PR TITLE
Adding support to populate node labels when GPUs are configured for passthrough

### PR DESCRIPTION
**Pull request by visheshtanksale**
Adding support to populate node labels when GPUs are configured for passthrough
GFD does not populate GPU related node labels in passthrough mode because it reads GPU details from NVML. This PR adds support to read GPU details from sysfs when GPU are configured for passthrough

To Do: Adding unit tests

### CODE REVIEW

File: api/config/v1/flags.go
Comment (klueska) [inline api/config/v1/flags.go:109]: I would drop the `Workload` and maybe rename the flag completely to something with `vfio` in the name

File: internal/resource/sysfs-lib.go
Comment (elezar) [inline internal/resource/sysfs-lib.go:45]: We may want to allow for injection of an `nvpci.Interface` for testing.

File: internal/resource/sysfs-lib.go
Comment (elezar) [inline internal/resource/sysfs-lib.go:57]: Since we `return` in the previous block, we don't need the explicit else.
```suggestion
		}
		vfioDev := vfioDevice{dev}
		devices = append(devices, vfioDev)
```

File: internal/resource/sysfs-lib.go
Comment (elezar) [inline internal/resource/sysfs-lib.go:53]: Question: Is this an error, or should we just skip the device?

File: internal/resource/factory.go
Comment (elezar) [inline internal/resource/factory.go:42]: See https://github.com/NVIDIA/k8s-device-plugin/pull/673 where I reworked this to use go-nvlib functionality that I'm trying to add.

If we add a `mode` flag, I would recommend  only passing the `mode` here and then making a construction decision based on that.

**Review by elezar — COMMENTED** [review]
Thanks @visheshtanksale. I think this is a good start.

I think adding a more generic mode will allow for more flexibility going forward, and also make it clear that NVML, CUDA and vfio labels are mutually exclusive.

File: api/config/v1/flags.go
Comment (elezar) [reply api/config/v1/flags.go:109]: @klueska as I suggested below, adding a top-level mode flag to switch between this, `nvml`, and `cuda` explicitly may be useful. Do you have specific thoughts on this?

File: internal/resource/factory.go
Comment (cdesiniotis) [inline internal/resource/factory.go]: I agree with Evan's comment: https://github.com/NVIDIA/k8s-device-plugin/pull/676/files#r1575766699

We should add a new `vfio` mode to the `go-nvlib/info` package, which represents a system where all GPUs are bound to the `vfio-pci` driver. We can construct a `VfioManager` if we detect a `vfio` system, so no need to introduce a new `VMPassthroughWorkload` flag.

File: internal/resource/sysfs-lib.go
Comment (visheshtanksale) [reply internal/resource/sysfs-lib.go]: I had return by mistake. Fixed it

File: internal/resource/sysfs-lib.go
Comment (visheshtanksale) [reply internal/resource/sysfs-lib.go]: Yes we should skip the device. Fixed it.

File: internal/resource/factory.go
Comment (visheshtanksale) [reply internal/resource/factory.go:42]: @elezar This makes sense. How do you want to handle this? Since the go-nvlib PR is still open do you want to update the info interface or should I wait for it to be merged?

Comment (elezar) [reply internal/resource/factory.go:42]: I think the addition of the mode flag is separate from that PR. We can already add the config option and command line flags here and then ensure that it is plumbed through. I would suggest.

1. Add a new PR that adds the mode flag / config option.
2. Rebase these changes to include the mode flag commits, adding support for a `vfio` mode.

Once the go-nvlib changes and #673 are merged, we can update the checks for the existing modes here.

As a follow-up we can pull the `vfio` detection into `go-nvlib`, but I don't see that as a blocker for this PR.

File: internal/resource/factory.go
Comment (elezar) [reply internal/resource/factory.go]: To clarify. The `mode` flag I suggested in the linked comment is local. We could (and probably should) pull some of the functionality into the `go-nvlib/info` package.

The important functionality from the perspective of GFD and the device plugin is that a user can explicitly override the mode if our autodetection fails. This already happens on newer Tegra-based systems, for example.

File: api/config/v1/flags.go
Comment (ArangoGutierrez) [inline api/config/v1/flags.go:166]: Could you add this to the Helm templates as well

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]

File: internal/resource/sysfs-lib.go
Comment (visheshtanksale) [reply internal/resource/sysfs-lib.go]: Fixed!

File: api/config/v1/flags.go
Comment (elezar) [inline api/config/v1/flags.go:109]: Question: Is this a top-level flag instead? That is to say, would this mode control both GFD and the device plugin?

File: cmd/gpu-feature-discovery/main.go
Comment (elezar) [inline cmd/gpu-feature-discovery/main.go:104]: `auto (default)`, `nvml`, `tegra`, `vfio`.

File: cmd/gpu-feature-discovery/main_test.go
Comment (elezar) [inline cmd/gpu-feature-discovery/main_test.go:217]: This should ideally default to `auto` -- although we can use `""` to mean this too.

File: internal/resource/factory.go
Comment (elezar) [inline internal/resource/factory.go:44]: This should accept a `string` and not a `*string`

File: cmd/gpu-feature-discovery/main.go
Comment (elezar) [reply cmd/gpu-feature-discovery/main.go:104]: We also want to add a check to `validateFlags` that checks for these valid values.

File: internal/resource/factory.go
Comment (elezar) [inline internal/resource/factory.go:76]: Let's just use `mode == "vfio"` here and require lower case input.

File: internal/resource/sysfs-lib.go
Comment (elezar) [inline internal/resource/sysfs-lib.go:30]: Remove the newline.

File: internal/resource/sysfs-lib.go
Comment (elezar) [inline internal/resource/sysfs-lib.go:31]: Do we need to pass the driver root here?

File: internal/resource/sysfs-lib.go
Comment (elezar) [inline internal/resource/sysfs-lib.go:25]: Do we really want to expose the entire `nvpci.Interface` here? can we use:
```
type vfioLib struct {
    nvpcilib nvpci.Interface
}
```
instead to ensure that we don't do so?

File: internal/resource/factory.go
Comment (elezar) [inline internal/resource/factory.go:71]: I would recommend first adding the mode flag so that we can properly handle the `nvml` and `tegra` modes and then rebase this PR onto those changes.

File: internal/resource/sysfs-device.go
Comment (elezar) [inline internal/resource/sysfs-device.go:26]: let's rather make this a private member.

**Review by elezar — COMMENTED** [review]
Thanks @visheshtanksale. I think this is moving in the right direction. It may be good to add the mode flag -- supporting the values `nvml` and `tegra` in a separate PR and then updating this to *add* the `vfio` mode. 

Apart from my implementation-specific coments, the one question I have is wheter the `mode` makes sense as a top-level (i.e. not GFD-specific) option.

File: internal/resource/factory.go
Comment (visheshtanksale) [reply internal/resource/factory.go]: Added this

File: api/config/v1/flags.go
Comment (visheshtanksale) [reply api/config/v1/flags.go:109]: Does running in the TEGRA mode mean anything to device plugin? Also currently we dont support NVML as well. So I feel it makes sense to keep it here.

File: cmd/gpu-feature-discovery/main_test.go
Comment (visheshtanksale) [reply cmd/gpu-feature-discovery/main_test.go:217]: Fixed

File: cmd/gpu-feature-discovery/main.go
Comment (visheshtanksale) [reply cmd/gpu-feature-discovery/main.go]: Fixed both!

File: internal/resource/factory.go
Comment (visheshtanksale) [reply internal/resource/factory.go]: Fixed

File: internal/resource/factory.go
Comment (visheshtanksale) [reply internal/resource/factory.go]: Fixed

File: internal/resource/sysfs-lib.go
Comment (visheshtanksale) [reply internal/resource/sysfs-lib.go]: Fixed

File: internal/resource/sysfs-lib.go
Comment (visheshtanksale) [reply internal/resource/sysfs-lib.go]: Fixed

File: internal/resource/sysfs-device.go
Comment (visheshtanksale) [reply internal/resource/sysfs-device.go]: Fixed

File: internal/resource/sysfs-lib.go
Comment (visheshtanksale) [reply internal/resource/sysfs-lib.go:30]: What do you mean by driver root?

Comment (elezar) [reply internal/resource/sysfs-lib.go:30]: I mean whether one of the `Options`:
```
// Option defines a function for passing options to the New() call.
type Option func(*nvpci)

// WithLogger provides an Option to set the logger for the library.
func WithLogger(logger logger) Option {
	return func(n *nvpci) {
		n.logger = logger
	}
}

// WithPCIDevicesRoot provides an Option to set the root path
// for PCI devices on the system.
func WithPCIDevicesRoot(root string) Option {
	return func(n *nvpci) {
		n.pciDevicesRoot = root
	}
}

// WithPCIDatabasePath provides an Option to set the path
// to the pciids database file.
func WithPCIDatabasePath(path string) Option {
	return func(n *nvpci) {
		n.pcidbPath = path
	}
}
```
that `nvpci.New()` accepts is required. I suppose since we're always running in a container we don't expect a different root for pci devices or a different pcidb path.

File: api/config/v1/flags.go
Comment (elezar) [reply api/config/v1/flags.go:109]: We have similar logic in the device plugin here: https://github.com/NVIDIA/k8s-device-plugin/blob/cf082174cda71b023030df58a6639572aaaea040/internal/plugin/manager/factory.go#L97-L133

This also changes the way resources are disovered based on autodetecting the platform (nvml, and tegra). I would assume that we would do something similar to allow the device plugin to also serve vfio devices?

Comment (elezar) [reply api/config/v1/flags.go:109]:
```suggestion
	Mode            *string   `json:"mode"            yaml:"mode"`
```

File: cmd/gpu-feature-discovery/main_test.go
Comment (elezar) [reply cmd/gpu-feature-discovery/main_test.go:217]: How did you fix this? I would have expected this to change to `ptr("auto")` instead.

Comment (visheshtanksale) [reply cmd/gpu-feature-discovery/main_test.go]: Apologies I missed this. Fixed it now.

File: internal/resource/sysfs-lib.go
Comment (visheshtanksale) [reply internal/resource/sysfs-lib.go:30]: Yes we dont expect both of the paths to be different, so not setting them.

File: api/config/v1/flags.go
Comment (elezar) [inline api/config/v1/flags.go:66]: Nit: Let's add this above `Plugin` to so that the plugin and gfd-specific flags are last.

File: internal/resource/factory.go
Comment (elezar) [inline internal/resource/factory.go:57]: nit: We may want to include `resolved` here:
```suggestion
	klog.Warningf("Unsupported mode detected: %v; using empty manager.", resolved)
```

File: internal/resource/sysfs-lib.go
Comment (elezar) [inline internal/resource/sysfs-lib.go:2]: nit: 2022 -> 2024

File: cmd/gpu-feature-discovery/main.go
Comment (elezar) [inline cmd/gpu-feature-discovery/main.go:107]:
```suggestion
			EnvVars: []string{"MODE", "GFD_MODE"},
```
or even just `MODE`?

**Review by elezar — CHANGES_REQUESTED** [review]
Thanks @visheshtanksale. I think this looks a lot better now. A minor comment on the ordering of the config structs. Once that's been addressed, I think we can move ahead.

File: internal/resource/factory.go
Comment (visheshtanksale) [reply internal/resource/factory.go]: Fixed

File: internal/resource/sysfs-lib.go
Comment (visheshtanksale) [reply internal/resource/sysfs-lib.go]: Fixed

File: cmd/gpu-feature-discovery/main.go
Comment (visheshtanksale) [reply cmd/gpu-feature-discovery/main.go]: Fixed

File: api/config/v1/flags.go
Comment (visheshtanksale) [reply api/config/v1/flags.go]: Done

**Review by ArangoGutierrez — APPROVED** [review]

**Comment by visheshtanksale** [comment]
> PR looks in good shape now, but could be squash the commits, there are 6 commits with the same title, and all for review comments, also edit PR description to provide some context, not a JIRA link.
> 
> Thanks

@ArangoGutierrez Squashed the commits and edited PR description.

**Review by ArangoGutierrez — APPROVED** [review]

File: deployments/helm/nvidia-device-plugin/values.yaml
Comment (elezar) [reply deployments/helm/nvidia-device-plugin/values.yaml:38]: Created #710

**Pull request merged**
